buildcontext/gitlookup.go

```diff
@@ -0,0 +1,179 @@
+package buildcontext
+
+import (
+	"fmt"
+	"net/url"
+	"regexp"
+	"strings"
+
+	"github.com/pkg/errors"
+)
+
+type gitMatcher struct {
+	name     string
+	re       *regexp.Regexp
+	user     string
+	suffix   string
+	protocol string
+	password string
+	keyScan  string
+}
+
+// GitLookup looksup gits
+type GitLookup struct {
+	matchers []*gitMatcher
+	catchAll *gitMatcher
+}
+
+// NewGitLookup creates new lookuper
+func NewGitLookup() *GitLookup {
+	matchers := []*gitMatcher{
+		{
+			name:     "github.com",
+			re:       regexp.MustCompile("github.com/[^/]+/[^/]+"),
+			user:     "git",
+			suffix:   ".git",
+			protocol: "ssh",
+			keyScan:  "github.com ssh-rsa AAAAB3NzaC1yc2EAAAABIwAAAQEAq2A7hRGmdnm9tUDbO9IDSwBK6TbQa+PXYPCPy6rbTrTtw7PHkccKrpp0yVhp5HdEIcKr6pLlVDBfOLX9QUsyCOV0wzfjIJNlGEYsdlLJizHhbn2mUjvSAHQqZETYP81eFzLQNnPHt4EVVUh7VfDESU84KezmD5QlWpXLmvU31/yMf+Se8xhHTvKSCZIFImWwoG6mbUoWf9nzpIoaSjB+weqqUUmpaaasXVal72J+UX2B+2RPW3RcT0eOzQgqlJL3RKrTJvdsjE3JEAvGq3lGHSZXy28G3skua2SmVi/w4yCE6gbODqnTWlg7+wC604ydGXA8VJiS5ap43JXiUFFAaQ==",
+		},
+		{
+			name:     "gitlab.com",
+			re:       regexp.MustCompile("gitlab.com/[^/]+/[^/]+"),
+			user:     "git",
+			suffix:   ".git",
+			protocol: "ssh",
+			keyScan:  "gitlab.com ssh-rsa AAAAB3NzaC1yc2EAAAADAQABAAABAQCsj2bNKTBSpIYDEGk9KxsGh3mySTRgMtXL583qmBpzeQ+jqCMRgBqB98u3z++J1sKlXHWfM9dyhSevkMwSbhoR8XIq/U0tCNyokEi/ueaBMCvbcTHhO7FcwzY92WK4Yt0aGROY5qX2UKSeOvuP4D6TPqKF1onrSzH9bx9XUf2lEdWT/ia1NEKjunUqu1xOB/StKDHMoX4/OKyIzuS0q/T1zOATthvasJFoPrAjkohTyaDUz2LN5JoH839hViyEG82yB+MjcFV5MU3N1l1QL3cVUCh93xSaua1N85qivl+siMkPGbO5xR/En4iEY6K2XPASUEMaieWVNTRCtJ4S8H+9",
+		},
+		{
+			name:     "bitbucket.com",
+			re:       regexp.MustCompile("bitbucket.com/[^/]+/[^/]+"),
+			user:     "git",
+			suffix:   ".git",
+			protocol: "ssh",
+			keyScan:  "bitbucket.com ssh-rsa AAAAB3NzaC1yc2EAAAABIwAAAQEAubiN81eDcafrgMeLzaFPsw2kNvEcqTKl/VqLat/MaB33pZy0y3rJZtnqwR2qOOvbwKZYKiEO1O6VqNEBxKvJJelCq0dTXWT5pbO2gDXC6h6QDXCaHo6pOHGPUy+YBaGQRGuSusMEASYiWunYN0vCAI8QaXnWMXNMdFP3jHAJH0eDsoiGnLPBlBp4TNm6rYI74nMzgz3B9IikW4WVK+dc8KZJZWYjAuORU3jc1c/NPskD2ASinf8v3xnfXeukU0sJ5N6m5E8VLjObPEO+mN2t/FZTMZLiFqPWc/ALSqnMnnhwrNi2rbfg/rd/IpL8Le3pSBne8+seeFVBoGqzHM9yXw==",
+		},
+	}
+
+	gl := &GitLookup{
+		matchers: matchers,
+		catchAll: &gitMatcher{
+			name:     "",
+			re:       regexp.MustCompile("[^/]+/[^/]+/[^/]+"),
+			user:     "git",
+			suffix:   ".git",
+			protocol: "ssh",
+		},
+	}
+	return gl
+}
+
+// ErrNoMatch occurs when no git matcher is found
+var ErrNoMatch = fmt.Errorf("no git match found")
+
+// DisableSSH changes all git matchers from ssh to https
+func (gl *GitLookup) DisableSSH() {
+	for i, m := range gl.matchers {
+		if m.protocol == "ssh" {
+			gl.matchers[i].protocol = "https"
+		}
+	}
+	if gl.catchAll.protocol == "ssh" {
+		gl.catchAll.protocol = "https"
+	}
+}
```

Comment:
So if no SSH agent is found or no keys are registered, we automatically switch to trying out https instead, right? Some questions:

1. If user does have valid SSH key, but their key is not registered with GitHub, is there a way in which we could try HTTPS in that situation?
2. What happens if the user's Earthly config for `github.com` says `auth: ssh`, but user doesn't have an SSH agent or no SSH key? Does the HTTPS fallback still kick in in this case?

Reply:
1) We only switch over to https when no agent is found, so unfortunately not; I think we would have to manually try a ssh to the host to see if it's accepted or not. Maybe we could try this in a future PR?

2) it will automatically get flipped back to https, so that might be confusing to a user. Maybe we should only auto-fall back when no auth is specified in the config?

Reply:
1. Yeah, that makes sense. I think that we should still keep instead-of in this case then - otherwise the experience of the user is less than what they have today for this specific edge case. Though this means that we have to implement the instead-of ourselves on the client side...
2. Yeah. Perhaps "auto" is the default and so if nothing is specified, "auto" allows us to flip to either.

Reply:
sounds great, just pushed up some new code for the auto mode.